igvm/src/c_api.rs

```diff
@@ -0,0 +1,366 @@
+        }
+        Error::InvalidBinaryDirectiveHeader(_) => {
+            IgvmApiError::IGVMAPI_INVALID_BINARY_DIRECTIVE_HEADER
+        }
+        Error::MultiplePlatformHeadersWithSameIsolation => {
+            IgvmApiError::IGVMAPI_MULTIPLE_PLATFORM_HEADERS_WITH_SAME_ISOLATION
+        }
+        Error::InvalidParameterAreaIndex => IgvmApiError::IGVMAPI_INVALID_PARAMETER_AREA_INDEX,
+        Error::InvalidPlatformType => IgvmApiError::IGVMAPI_INVALID_PLATFORM_TYPE,
+        Error::NoFreeCompatibilityMasks => IgvmApiError::IGVMAPI_NO_FREE_COMPATIBILITY_MASKS,
+        Error::InvalidFixedHeader => IgvmApiError::IGVMAPI_INVALID_FIXED_HEADER,
+        Error::InvalidBinaryVariableHeaderSection => {
+            IgvmApiError::IGVMAPI_INVALID_BINARY_VARIABLE_HEADER_SECTION
+        }
+        Error::InvalidChecksum {
+            expected: _,
+            header_value: _,
+        } => IgvmApiError::IGVMAPI_INVALID_CHECKSUM,
+        Error::MultiplePageTableRelocationHeaders => {
+            IgvmApiError::IGVMAPI_MULTIPLE_PAGE_TABLE_RELOCATION_HEADERS
+        }
+        Error::RelocationRegionsOverlap => IgvmApiError::IGVMAPI_RELOCATION_REGIONS_OVERLAP,
+        Error::ParameterInsertInsidePageTableRegion => {
+            IgvmApiError::IGVMAPI_PARAMETER_INSERT_INSIDE_PAGE_TABLE_REGION
+        }
+        Error::NoMatchingVpContext => IgvmApiError::IGVMAPI_NO_MATCHING_VP_CONTEXT,
+        Error::PlatformArchUnsupported {
+            arch: _,
+            platform: _,
+        } => IgvmApiError::IGVMAPI_PLATFORM_ARCH_UNSUPPORTED,
+        Error::InvalidHeaderArch {
+            arch: _,
+            header_type: _,
+        } => IgvmApiError::IGVMAPI_INVALID_HEADER_ARCH,
+        Error::UnsupportedPageSize(_) => IgvmApiError::IGVMAPI_UNSUPPORTED_PAGE_SIZE,
+        Error::InvalidFixedHeaderArch(_) => IgvmApiError::IGVMAPI_INVALID_FIXED_HEADER_ARCH,
+        Error::MergeRevision => IgvmApiError::IGVMAPI_MERGE_REVISION,
+    }
+}
+
+fn copy_header_data(src: &Vec<u8>, dst: *mut u8, dst_len: *mut u32) -> IgvmApiError {
+    // Safety: Using pointer from C
+    unsafe {
+        if src.len() > *dst_len as usize {
+            *dst_len = src.len() as u32;
+            return IgvmApiError::IGVMAPI_BUFFER_TOO_SMALL;
+        }
+
+        core::ptr::copy(src.as_ptr(), dst, src.len());
+        *dst_len = src.len() as u32;
+    }
+    IgvmApiError::IGVMAPI_OK
+}
+
+/// Parse a binary array containing an IGVM file.
+/// The contents of the file are validated and, if valid, a handle is returned
+/// to represent the file. This handle must be freed with a call to
+/// igvm_free().
+///
+/// # Safety
+/// This function is exported and designed to be called as a C API. As such
+/// it works with raw pointers and is inherently unsafe due to this.
+#[no_mangle]
+pub unsafe extern "C" fn igvm_new_from_binary(
+    data: *const u8,
+    len: u32,
+    igvm: *mut *mut IgvmFile,
+) -> IgvmApiError {
+    // Safety: data array must be at least as large as len
```

Comment:
nit: Can we do `SAFETY:` instead of `Safety:`. Also, we should comment exactly what the preconditions the caller must adhere to, in this case that data must be as large as len in the doc comment. Then, here in the function safety, we say "The caller guarantees that data is valid for len bytes".

Reply:
As what we really want with Safety docs here (and safety comments) is to make it clear _why_ the function is unsafe. 

And for unsafe blocks, what conditions must hold, and why this unsafe call is valid.

Reply:
My bad. I put `Safety:` placeholders there to prevent the clippy warnings then forgot to finish the documentation before creating this PR. I've now updated the docs and switched to `SAFETY:`.